### PR TITLE
#438 - Added KaldiSTT as a Speech to Text-Engine.

### PR DESCRIPTION
**Pull request by nilsseipel**
Kaldi gstreamer server has to run on localhost. Use this kaldi gstreamer server: https://github.com/alumae/kaldi-gstreamer-server 
Tested on a local machine with the kaldigstserver shipped with lucida (another AI). For using KaldiSTT you have to set the config for server to "update": false and the stt to "module": "kaldi" with "port" to your kaldigstserver port Number. Default is 8081.

### CODE REVIEW

File: mycroft/stt/__init__.py [hunk 1]
```diff
@@ -101,13 +107,38 @@ def execute(self, audio, language=None):
...
+    def execute(self, audio, language=None):
+        language = language or self.lang
+        
+        port_kaldi = 8081 #default port        
+        config = ConfigurationManager.get().get("stt", {})
+        port_kaldi = config.get("port")
+        print("Port for localhost Kaldi Server:", port_kaldi)
+        
+        #the kaldigstserver has to run at localhost on the defined port in the config file under stt "port": xxxx
+        kaldi_server_response = requests.post("http://localhost:%s/client/dynamic/recognize" % port_kaldi, data=audio.get_wav_data())
```
Comment (jdorleans) [inline mycroft/stt/__init__.py]: As suggested above, use an URI in your kalfi configuration and you don't have to bind to a fix host and port.

Then, you'd have:
```
response = requests.post(config.get("uri"), data=audio.get_wav_data())        
```

NOTE: This lonely line would replace your code from line 120 to 127.

File: mycroft/stt/__init__.py [hunk 2]
```diff
@@ -101,13 +107,38 @@ def execute(self, audio, language=None):
...
+        port_kaldi = config.get("port")
+        print("Port for localhost Kaldi Server:", port_kaldi)
+        
+        #the kaldigstserver has to run at localhost on the defined port in the config file under stt "port": xxxx
+        kaldi_server_response = requests.post("http://localhost:%s/client/dynamic/recognize" % port_kaldi, data=audio.get_wav_data())        
+        kaldi_json = json.loads(kaldi_server_response.text)
+        hypotheses = kaldi_json["hypotheses"]
+        
+        #re.sub... deletes all [noise] inserts made by kaldi        
+        return re.sub(r'\s*\[noise\]\s*', '', hypotheses[0]["utterance"])
```
Comment (jdorleans) [inline mycroft/stt/__init__.py]: Finally, from 128 to 132:
```
return get_response(response)

def get_response(self, response):
  try:
    hypotheses = response.json()["hypotheses"]
    return re.sub(r'\s*\[noise\]\s*', '', hypotheses[0]["utterance"])
  except:
    return None
```

File: mycroft/stt/__init__.py [hunk 3]
```diff
@@ -22,6 +22,12 @@
 from mycroft.configuration import ConfigurationManager
 from mycroft.util.log import getLogger
 
+
+#imports for json handling, requests and string formatting for KaldiSTT
```
Comment (jdorleans) [inline mycroft/stt/__init__.py]: no need for comments for imports... btw, you probably will revisit those with the later changes.

Comment (nilsseipel) [inline mycroft/stt/__init__.py]: I've implemented all changes and it's working fine. But I don't know how to rebase with dev and doesn't have the time to find out now. Here is my work from now: 

`
class KaldiSTT(STT):
    def __init__(self):
        super(KaldiSTT, self).__init__()

    def execute(self, audio, language=None):
        language = language or self.lang
          
        config = ConfigurationManager.get().get("stt", {})
       
        response = requests.post(config.get("uri"), data=audio.get_wav_data())        
        return get_response(response)

def get_response(response):
    try:
      hypotheses = response.json()["hypotheses"]
      return re.sub(r'\s*\[noise\]\s*', '', hypotheses[0]["utterance"])
    except:
      return None
        `

Comment (jdorleans) [inline mycroft/stt/__init__.py]: Sorry about the confusion. The configuration should use the `self` operator. That way, you'll have access to the module configuration itself. Please, do the following update:
```
    def execute(self, audio, language=None):
        language = language or self.lang
        response = requests.post(self.config.get("uri"), data=audio.get_wav_data())        
        return get_response(response)
```
Then, your config should be like: 
```
  "stt": {
    "module": "kaldi",
    "kaldi": {
      "uri": "http://localhost:8081/client/dynamic/recognize"
    }
  }
```
To rebase you just have to update your local `dev` branch based on our `upstream`. Then, rebase your feature on it:
```
# make sure you have your upstream setup
git remote add upstream git@github.com:MycroftAI/mycroft-core.git

# update your local dev branch based on our upstream
git fetch upstream
git checkout dev
git reset --hard upstream/dev

# rebase your feature branch based on dev and force the push to remote
git checkout feature/issues-438
git rebase dev
git push -f
```

**Comment by Khazrak** [comment]
Any status on this?

**Comment by aatchison** [comment]
I would like someone to review this. It's taking forever, sorry.

**Comment by Teagan42** [comment]
👍 bump

Looks like there are some conflicts and unaddressed comments, but damn would this be a nice addition

File: mycroft/stt/__init__.py (hunk 1 above)
Comment (Teagan42) [inline mycroft/stt/__init__.py]: I agree - this code forces the user to have kaldi running on the same box as Mycroft - when in reality that will probably be an edge case. I have Kaldi running on my home-automation server, and would then point any mycroft instances at the Kaldi server

**Comment by Teagan42** [comment]
@nilsseipel if you don't mind, I'm willing to take your changes and address the concerns from @jdorleans  this weekend in a separate PR.

**Comment by ethanaward** [comment]
@nilsseipel @Teagan42 Has there been any progress on an updated version of this?

**Comment by Teagan42** [comment]
Sorry I got side tracked in other things. Will try to look into it this
weekend

**Comment by augustnmonteiro** [comment]
@nilsseipel can you rebase this pr please. `git rebase dev`

**Comment by feitingen** [comment]
Sorry for being impatient, but is anyone working on it? 

I just tested it now, it's working really well on my computer, though I'd really prefer to configure the whole uri instead of just the port.

**Comment by Teagan42** [comment]
Created a new PR with the changes w/ the requested code changes above: https://github.com/MycroftAI/mycroft-core/pull/878 @feitingen can you please test my branch - I haven't stood kaldi back up since I wiped my server the other day.

**Comment by Teagan42** [comment]
This can be closed as the other PR completed this work

**Comment by forslund** [comment]
Implemented in #878. Closing.

**Pull request closed without merging**